### PR TITLE
 leakyRelu0.5^2 + GPTQ + EMA + BigramHash(1.3069)

### PR DESCRIPTION
**Pull request by Hieuabssy**

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review —  leakyRelu0.5^2 + GPTQ + EMA + BigramHash(1.3069)

**Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

## Analysis **Head SHA:** d5b6712b0248f367b00a4c0847362f155ac5e449 **Files changed:** log/README.md, log/submission.json, log/train_gpt.py, log/train_seed_{1337,2026,42}.txt, train_gpt.py --- ### 1. ILLEGAL n-gram family bug check — CLEAR The BigramHash implementation (lines 610–636) uses the standard (prev_token, cur_token) bigram hash: ``` bigram_idx = (prev.long() * 1_000_003 + input_ids.long()) % self.bigram_vocab_size ``` `prev` is `input_ids[:, :-1]` shifted right — purely causal, from the input sequence. The target tokens (`y_tok`, `target_ids`) are never incorporated into the hash key at any point. The `BigramHashEmbedding.forward` takes only `input_ids` (line 625), and the model's `forward` (line 708) calls `self.bigram_emb(input_ids)` before seeing targets. No XOR or additive leakage of labels into the embedding key. **NOT a family bug.** ### 2. ILLEGAL Pre-Quant TTT check — CLEAR No test-time training loop exists. `eval_val` (lines 225–291) runs entirely under `torch.inference_mode()` and `model.eval()`. No backward pass, no optimizer step, no gradient accumulation is invoked during evaluation. The function is called twice at `last_step` (lines 997 and 1046) — once for the live model, once for the int8 roundtrip model — but both are purely inference. **No TTT of any kind.** ### 3. Score-first TTT / is_last_chunk guard — NOT PRESENT No TTT pattern at all; this check is not applicable. ### 4. Scored-region SLOT — NOT PRESENT No held scored-region slot mechanism detected. ### 5. Overall architecture Standard causal LM with: U-Net skip connections, BigramHash embeddings (2048×6144), Muon optimizer for matrix params, EMA, QAT in the last 15% of training (STE int6 fake-quant via `CastedLinear._qat_enabled`), post-training GPTQ-lite int8 + zlib compression. No val-data leakage, no...

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the audit — this looks like a clean pure-neural submission.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*
